src/SonarScanner.MSBuild.PreProcessor/SonarQubeServer.cs

```diff
@@ -0,0 +1,136 @@
+            this.logger = logger;
+        }
+
+        public IList<SonarQubeRule> GetActiveRules(string qprofile) =>
+            EnsureConnected(
+                () => sonarQubeService.GetRulesAsync(true, qprofile, CancellationToken.None));
+
+        public IEnumerable<string> GetAllLanguages() =>
+            EnsureConnected(
+                () => sonarQubeService.GetAllLanguagesAsync(CancellationToken.None)).Select(l => l.Key);
+
+        public IList<SonarQubeRule> GetInactiveRules(string qprofile, string language) =>
+            EnsureConnected(
+                () => sonarQubeService.GetRulesAsync(false, qprofile, CancellationToken.None));
+
+        public IDictionary<string, string> GetProperties(string projectKey, string projectBranch)
+        {
+            var properties = EnsureConnected(
+                () => sonarQubeService.GetAllPropertiesAsync(GetCompositeProjectKey(projectKey, projectBranch), CancellationToken.None))
+                .ToDictionary(p => p.Key, p => p.Value);
+
+            ValidateTestProjectPattern(properties);
+
+            return properties;
+        }
+
+        private static string GetCompositeProjectKey(string projectKey, string projectBranch) =>
+            string.IsNullOrEmpty(projectBranch)
+                ? projectKey
+                : $"{projectKey}:{projectBranch}";
+
+        public Version GetServerVersion() =>
+            sonarQubeService.SonarQubeVersion;
+
+        public bool TryDownloadEmbeddedFile(string pluginKey, string embeddedFileName, string targetDirectory)
+        {
+            var stream = EnsureConnected(
+                () => sonarQubeService.DownloadStaticFileAsync(pluginKey, embeddedFileName, CancellationToken.None));
+
+            using (var file = File.Create(Path.Combine(targetDirectory, embeddedFileName)))
+            {
+                stream.CopyTo(file);
+                return true;
+            }
+        }
+
+        public bool TryGetQualityProfile(string projectKey, string projectBranch, string organization,
+            string language, out string qualityProfileKey)
+        {
+            var qualityProfile = EnsureConnected(
+                () => sonarQubeService.GetQualityProfileAsync(
+                    GetCompositeProjectKey(projectKey, projectBranch),
+                    organization,
+                    new SonarQubeLanguage(language, language),
+                    CancellationToken.None));
+
+            qualityProfileKey = qualityProfile.Key;
+            return true;
+        }
+
+        private T EnsureConnected<T>(Func<Task<T>> func)
+        {
+            if (!isConnected)
+            {
+                sonarQubeService
+                    .ConnectAsync(connectionInformation, CancellationToken.None)
+                    .GetAwaiter()
+                    .GetResult();
+                isConnected = true;
+            }
+            return func().GetAwaiter().GetResult();
```

Comment:
This method should be ok for single-threaded application, but consider developing more robust mechanism for synchronized isConnected checks.